app/src/main/java/org/schabi/newpipe/util/potoken/PoTokenWebView.kt

```diff
@@ -164,46 +158,41 @@ class PoTokenWebView private constructor(
             val (integrityToken, expirationTimeInSeconds) = parseIntegrityTokenData(responseBody)
 
             // leave 10 minutes of margin just to be sure
-            expirationInstant = Instant.now().plusSeconds(expirationTimeInSeconds - 600)
+            expirationInstant = Instant.now().plusSeconds(expirationTimeInSeconds).minus(10, ChronoUnit.MINUTES)
 
-            webView.evaluateJavascript(
-                "this.integrityToken = $integrityToken"
-            ) {
+            webView.evaluateJavascript("this.integrityToken = $integrityToken") {
                 if (BuildConfig.DEBUG) {
                     Log.d(TAG, "initialization finished, expiration=${expirationTimeInSeconds}s")
                 }
-                generatorEmitter.onSuccess(this)
+                continuation.resume(this)
             }
         }
     }
     //endregion
 
     //region Obtaining poTokens
-    override fun generatePoToken(identifier: String): Single<String> =
-        Single.create { emitter ->
-            if (BuildConfig.DEBUG) {
-                Log.d(TAG, "generatePoToken() called with identifier $identifier")
-            }
-            runOnMainThread(emitter) {
-                addPoTokenEmitter(identifier, emitter)
-                val u8Identifier = stringToU8(identifier)
+    override suspend fun generatePoToken(identifier: String): String {
+        return withContext(Dispatchers.Main) {
+            suspendCancellableCoroutine { cont ->
+                if (BuildConfig.DEBUG) {
+                    Log.d(TAG, "generatePoToken() called with identifier $identifier")
+                }
+                addPoTokenEmitter(identifier, cont)
                 webView.evaluateJavascript(
                     """try {
                         identifier = "$identifier"
-                        u8Identifier = $u8Identifier
+                        u8Identifier = ${stringToU8(identifier)}
                         poTokenU8 = obtainPoToken(webPoSignalOutput, integrityToken, u8Identifier)
-                        poTokenU8String = ""
-                        for (i = 0; i < poTokenU8.length; i++) {
-                            if (i != 0) poTokenU8String += ","
-                            poTokenU8String += poTokenU8[i]
-                        }
+                        poTokenU8String = poTokenU8.join(",")
                         $JS_INTERFACE.onObtainPoTokenResult(identifier, poTokenU8String)
                     } catch (error) {
                         $JS_INTERFACE.onObtainPoTokenError(identifier, error + "\n" + error.stack)
                     }""",
```

Comment:
Do not change the javascript in any place. This needs to work on browsers from 20 years ago, which some devices still use as webviews. If you check the PR adding po token support, I specifically worked on making the javascript as old as possible to avoid any incompatibilities, and neither `.join()` nor `stringToU8()` are good.